aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1338,8 +1336,53 @@ TORCH_IMPL_FUNC(mean_out)
       }
     }
     auto& result_mut = const_cast<Tensor&>(result);
-    at::sum_out(result_mut, self, opt_dim, keepdim, dtype).div_(dim_prod);
+    switch(dtype) {
+      case kHalf:
+      case kBFloat16:
+        // For accuracy reasons, BF16/FP16 mean should be computed via the
+        // following approach:
+        //  cast_fp32 -> sum -> div -> cast_bf16_or_fp16
+        //
+        // Such an approach is necessary because if we were to choose the same
+        // approach for BF16/FP16 as FP32 here, then it would have resulted in
+        // the following code-flow -
+        // cast_fp32 -> sum -> cast_bf16 -> cast_fp32 -> div -> cast_bf16,
+        // which, in turn, does not produce as accurate results.
+        result_mut = result_mut.to(ScalarType::Float,
+                                   result_mut.layout(),
+                                   result_mut.device(),
+                                   /*pin_memory=*/false,
+                                   /*non_blocking=*/false,
+                                   /*copy=*/false,
+                                   /*memory_format=*/at::MemoryFormat::Preserve);
+
+        // self (input tensor) will initially be cast to FP32 in sum_out.
+        // This results in having to read that FP32 tensor again, but maybe in
+        // the future, we could revise the implementation to not materialize
+        // that intermediate FP32 tensor. That approach would probably require
+        // some modifications in binary_kernel_reduce_vec(),
+        // TensorIteratorBase::for_each(), and
+        // TensorIteratorBase::serial_for_each().
+        at::sum_out(
+            result_mut, self, opt_dim, keepdim, ScalarType::Float).div_(dim_prod);
```

Comment:
I guess you can share this line with fp32 path?

Reply:
No, because FP32 path is shared by some other dtypes as well. 
BTW, `mean_kernel_impl` is not being used by any dtype now.
Would you recommend removing it from the PyTorch codebase? Thanks!

Reply:
Eliminated duplicate code. Thank you!